### DISCUSSION TITLE
`from_integer` and `from_string` using `std::optional`

### DISCUSSION THREAD

**Discussion by raphaelzulliger** [actor-framework/actor-framework#1271] [Ideas]
While reading the docs, I stumbled across code like this:

```c++
bool from_string(caf::string_view in, math_error& out) {
  if (in == "division_by_zero") {
    out = math_error::division_by_zero;
    return true;
  } else {
    return false;
  }
}

bool from_integer(uint8_t in, math_error& out) {
  if (in == 1) {
    out = math_error::division_by_zero;
    return true;
  } else {
    return false;
  }
}
```

I expect this code was written before `std::optional` was available? Anyway, for a next version of CAF, one that is "allowed to" break the API, using `std::optional` could probably be considered for handling this. E.g. like this:

```c++
std::optional<bool> from_integer(uint8_t in) {
  if (in == 1) {
    return math_error::division_by_zero;
  } else {
    return {};
  }
}
```
(I didn't compile this - it may not work as is)

**Comment by Neverlord** [comment]
CAF has its own `optional` implementation, that's not the reason. I'm not particularly fond of output parameters either, but I think in this case they provide the best tradeoff. Suppose you are using `from_integer` for `math_error`. Then how would you go about adding another error category? C++ cannot overload functions on the return type, so one cannot have two `from_integer` functions both taking an `uint8_t` but return different types.

Using output parameters is the least "noisy" way to have a common function name for the conversions that only depends on ADL. This convention allows CAF to provide a default implementation where users can opt-in with little hassle, see https://github.com/actor-framework/actor-framework/blob/master/libcaf_core/caf/default_enum_inspect.hpp.

Having `from_integer` as a static member function would solve this more elegantly, but an `enum` cannot have member functions in C++. The next best thing would be providing a template (class or function) that users then specialize per type. That would work, but it would force users to add to the `caf` namespace and would involve more boilerplate code than the current free function interface.

**Reply by raphaelzulliger** [reply]
Thanks for the detailed explanation. I didn't think that far (I merely saw the code and thought: Oh, that's a perfect case for `std::optional`). Thereby I missed the bigger picture.

Supporting overloads must probably be done like that (unless, probably, some sophisticated template stuff)
